x-pack/libbeat/management/managerV2.go

```diff
@@ -311,95 +377,156 @@ func (cm *BeatV2Manager) stopBeat() {
-	cm.addUnit(unit)
-	unitType := unit.Type()
+	// reload the output configuration
+	var errs multierror.Errors
+	if err := cm.reloadOutput(outputUnit); err != nil {
+		errs = append(errs, err)
+	}
 
-	if unitType == client.UnitTypeOutput {
-		cm.handleOutputReload(unit)
-	} else if unitType == client.UnitTypeInput {
-		cm.handleInputReload(unit)
+	// compute the input configuration
+	//
+	// in v2 only a single input type will be started per component, so we don't need to
+	// worry about getting multiple re-loaders (we just need the one for the type)
+	if err := cm.reloadInputs(inputUnits); err != nil {
+		errs = append(errs, err)
 	}
-}
 
-// Handle the updated config for an output unit
-func (cm *BeatV2Manager) handleOutputReload(unit *client.Unit) {
-	_, _, rawConfig := unit.Expected()
-	if rawConfig == nil {
-		cm.logger.Infof("got output update with no config, ignoring")
-		return
+	// report the stopping units as stopped
+	for _, unit := range stoppingUnits {
+		_ = unit.UpdateState(client.UnitStateStopped, "Stopped", nil)
 	}
-	cm.logger.Debugf("Got Output unit config '%s'", rawConfig.GetId())
 
-	reloadConfig, err := groupByOutputs(rawConfig)
-	if err != nil {
-		errString := fmt.Errorf("Failed to generate config for output: %w", err)
-		_ = unit.UpdateState(client.UnitStateFailed, errString.Error(), nil)
-		return
+	// any error during reload changes the whole state of the beat to failed
+	if len(errs) > 0 {
+		cm.status = lbmanagement.Failed
+		cm.message = fmt.Sprintf("%s", errs)
+	}
+
+	// now update the statuses of all units
+	cm.mx.Lock()
+	status := getUnitState(cm.status)
+	message := cm.message
+	payload := cm.payload
+	cm.mx.Unlock()
+	for _, unit := range units {
+		state, _, _ := unit.Expected()
+		if state == client.UnitStateStopped {
+			// unit is expected to be stopping (don't adjust the state as the state is now managed by the
+			// `reload` method and will be marked stopped in that code path)
+			continue
+		}
+		err := unit.UpdateState(status, message, payload)
+		if err != nil {
+			cm.logger.Errorf("Failed to update unit %s status: %s", unit.ID(), err)
+		}
 	}
+}
+
+func (cm *BeatV2Manager) reloadOutput(unit *client.Unit) error {
 	// Assuming that the output reloadable isn't a list, see createBeater() in cmd/instance/beat.go
 	output := cm.registry.GetReloadableOutput()
 	if output == nil {
-		_ = unit.UpdateState(client.UnitStateFailed, "failed to find beat reloadable type 'output'", nil)
-		return
+		return fmt.Errorf("failed to find beat reloadable type 'output'")
 	}
 
-	_ = unit.UpdateState(client.UnitStateConfiguring, "reloading output component", nil)
-	cm.lastOutputCfg = reloadConfig
-	err = output.Reload(reloadConfig)
-	if err != nil {
-		errString := fmt.Errorf("Failed to reload component: %w", err)
-		_ = unit.UpdateState(client.UnitStateFailed, errString.Error(), nil)
-		return
+	if unit == nil {
+		// output is being stopped
+		err := output.Reload(nil)
```

Comment:
Is there any reason why this change doesn't also trigger the Beat to restart? I'm not sure this will be any better behaved, in fact I'm not sure of the behaviour of removing the output in a Beat in general.

Reply:
Because if the output is now nil, that means beats is stopping. We would not want it to just be stopped here, once all the units have reported stopped the Elastic Agent will stop the beat.

Reply:
Makes sense, probably worth adding a clarify this.